ci/docker-publish.yml

```diff
@@ -55,9 +68,35 @@ jobs:
       # Build and push Docker image with Buildx (don't push on PR)
       # https://github.com/docker/build-push-action
       - name: Build and push Docker image
+        id: build-and-push
         uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
         with:
           context: .
           push: ${{ github.event_name != 'pull_request' }}
           tags: ${{ steps.meta.outputs.tags }}
           labels: ${{ steps.meta.outputs.labels }}
+
+      # Check whether this is a public repository before signing the image.
+      # The keyless signing process records signatures on the Rekor public
+      # transparency log, so signing is disabled for private repos by default
+      # to avoid leaking private data.  If you wish to sign things anyways,
+      # then this check can be removed and --force can be added to the cosign
+      # command below.
+      - name: Confirm this is a public repository
+        if: ${{ github.event_name != 'pull_request' }}
+        env:
+          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+        run: |
+          TEMP=$(gh api repos/$GITHUB_REPOSITORY --jq '.private')
+          echo "IS_PRIVATE=$TEMP" >> $GITHUB_ENV
+
+      # Sign the resulting Docker image digest except on PRs and private repos
+      # https://github.com/sigstore/cosign
+      - name: Sign the published Docker image
+        if: ${{ github.event_name != 'pull_request' && env.IS_PRIVATE == 'false' }}
```

Comment:
I came here to suggest that we only run this on public repos, but you beat me to it!

I think we could simplify the implementation a bit though, by not having the above step and just checking `if: ${{ !github.event.repository.private }}` here.

It shouldn't be the default, but there might be use-cases where a private repo wants to use keyless signing, and is okay with the public record, so we could have an opt-in env variable (`COSIGN_USE_PUBLIC_KEYLESS_SIGNING`?) that uses keyless signing regardless. Not required, just a suggestion.

Thanks for contributing this functionality!

Reply:
@steiza even better, I'll update the PR.  The next version of `cosign` will make it so we can always sign and for private repos they would add `--force` if they want to publish to the transparency log, which feels like exactly the opt-in we want.

Reply:
Perfect, this is much tighter, and I just tried it on my public and private demo repos and it works like a charm.  ✨ 

Thanks again @steiza !